pytential/symbolic/compiler.py

```diff
@@ -569,20 +583,19 @@ def map_int_g(self, expr, name_hint=None):
         except KeyError:
             # make sure operator assignments stand alone and don't get muddled
             # up in vector arithmetic
-            density_var = self.assign_to_new_var(self.rec(expr.density))
+            density_vars = [self.assign_to_new_var(self.rec(density)) for
+                density in expr.densities]
 
             group = self.group_to_operators[self.op_group_features(expr)]
             names = [self.get_var_name() for op in group]
 
-            kernels = sorted({op.kernel for op in group}, key=repr)
-
-            kernel_to_index = {kernel: i for i, kernel in enumerate(kernels)}
-
+            sorted_ops = sorted(group, key=lambda op: repr(op.target_kernel))
```

Comment:
Wouldn't you need similar sorting/normalization logic for "source kernels"?

Reply:
target kernels sorting was probably needed to get deterministic code after collecting IntGs. Since source_kernels is an attribute of the IntGs, they need not be sorted. We could sort them along with the densities with the same ordering, but I don't think it's necessary.

Reply:
No, the sorting is there to standardize the operations. Otherwise, `S, dx S` and `dx S, S` would each generate their own code. IMO, a similar concern exists for source derivatives.

Btw, it might be useful to add a comment to the sort to explain its reason for existing.

Reply:
Got it. Then there's no need to sort target kernels then. Only source kernels.

Reply:
Not sure I agree. We should talk through the implications of this tomorrow.